rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala

```diff
@@ -1013,6 +1013,129 @@ object ProcNormalizeMatcher {
+                      ParLocallyFree.connectiveUsed(targetPar) || targetParsConnectiveUsed
+                    )
+                }
+            }
+
+        def normalizePatterns(
+            patternProcs: Vector[Proc]
+        ): M[Vector[(Par, DebruijnLevelMap[VarSort], BitSet)]] =
+          patternProcs.traverse { patternProc =>
+            normalizeMatch[M](
+              patternProc,
+              ProcVisitInputs(VectorPar(), input.env.pushDown(), DebruijnLevelMap[VarSort]())
+            ).map {
+              case ProcVisitOutputs(patternPar, patternParKnownFree) =>
+                (
+                  patternPar,
+                  patternParKnownFree,
+                  ParLocallyFree.locallyFree(patternPar, input.env.depth + 1)
+                )
+            }
+          }
+
+        def mergeFree(
+            knownFreeList: List[DebruijnLevelMap[VarSort]]
+        ): M[DebruijnLevelMap[VarSort]] =
+          knownFreeList.foldM(DebruijnLevelMap[VarSort]()) {
+            case (acc, knownFree) =>
+              acc.merge(knownFree) match {
+                case (declarationsKnownFree1, Nil) =>
+                  declarationsKnownFree1.pure[M]
+                case (_, (shadowingVariable, line, col) :: _) =>
+                  val Some((_, _, firstUsageLine, firstUsageCol)) =
+                    acc.get(shadowingVariable)
+                  UnexpectedReuseOfProcContextFree(
+                    shadowingVariable,
+                    firstUsageLine,
+                    firstUsageCol,
+                    line,
+                    col
+                  ).raiseError[M, DebruijnLevelMap[VarSort]]
+              }
+          }
+
+        /**
+          * Like binds and contracts, a variable can only be used once in a pattern per let expression.
+          *
+          */
+        for {
+          liftedDeclarations                                                                 <- p.listdecl_.toVector.traverse(liftDeclaration)
+          (patternProcs, targetProcs)                                                        = liftedDeclarations.unzip
+          normalizedTargets                                                                  <- normalizeTargets(targetProcs)
+          (targetPars, targetParsKnownFree, targetParsLocallyFree, targetParsConnectiveUsed) = normalizedTargets
+          normalizedPatterns                                                                 <- normalizePatterns(patternProcs)
+          (patternPars, patternParKnownFreeList, patternParLocallyFreeList)                  = normalizedPatterns.unzip3
+          patternParsLocallyFree                                                             = patternParLocallyFreeList.foldLeft(BitSet())(_ | _)
+          declarationsWithLocallyFree = patternPars
+            .zip(targetPars.reverse)
+            .zip(patternParKnownFreeList)
+          sortedDeclarationsWithLocallyFree <- DeclarationsSortMatcher
```

Comment:
Is there a need to sort declarations in the normalization step?

Reply:
Yes. We have to pre-sort declarations in the same way that we pre-sort receive bindings.